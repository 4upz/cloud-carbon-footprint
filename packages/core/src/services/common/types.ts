/*
 * Â© 2021 Thoughtworks, Inc. All rights reserved.
 */

export type QUERY_DATE_TYPES = {
  [key: string]: string
}
