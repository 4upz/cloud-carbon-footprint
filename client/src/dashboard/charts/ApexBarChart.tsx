/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent } from 'react'
import { makeStyles, useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { sumCO2, sumCO2ByServiceOrRegion } from '../transformData'
import { ApexChartProps } from './common/ChartTypes'

const useStyles = makeStyles(() => {
  return {
    root: {
      minHeight: '500px',
      maxHeight: '500px',
      overflowY: 'auto',
    },
  }
})

export const ApexBarChart: FunctionComponent<ApexChartProps> = ({ data, dataType }) => {
  const theme = useTheme()
  const classes = useStyles()
  const chartColors = [theme.palette.primary.main]
  const barChartData = sumCO2ByServiceOrRegion(data, dataType)

  const dataEntries: { x: string; y: number }[] = Object.entries(barChartData)
    .filter((item) => item[1] > 0)
    .map((item) => ({
      x: item[0],
      y: item[1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  // Added to dynamically resize height based on visible accounts/regions/services
  // noted that when data entries were minimal, the bars where extremely thin
  const determineHeight: () => number = () => {
    if (dataEntries.length === 1) {
      return dataEntries.length * 90
    } else if (dataEntries.length < 5) {
      return dataEntries.length * 70
    }

    return dataEntries.length * 55
  }

  const options = {
    series: [
      {
        name: 'Total CO2e',
        data: dataEntries,
      },
    ],
    colors: chartColors,
    chart: {
      type: 'bar',
      toolbar: {
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
      },
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (value: number) {
        return `${((value / sumCO2(data)) * 100).toFixed(2)} %`
      },
      offsetX: 10,
      background: {
        enabled: true,
        foreColor: theme.palette.primary.main,
        borderColor: theme.palette.primary.dark,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        show: false,
      },
      axisBorder: {
        show: false,
      },
    },
    yaxis: {
      labels: {
        style: {
          fontSize: '13px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      x: {
        show: false,
      },
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} mt`
        },
      },
    },
    height: determineHeight(),
  }

  return (
    <div className={classes.root}>
      <Chart options={options} series={options.series} type="bar" height={options.height} />
    </div>
  )
}
