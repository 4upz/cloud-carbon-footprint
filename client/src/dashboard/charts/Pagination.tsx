/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { useState, useEffect, PropsWithChildren, ReactElement } from 'react'
import { ChevronLeft, ChevronRight } from '@material-ui/icons'
import { IconButton, makeStyles } from '@material-ui/core'

interface UsePaginateData<T> {
  paginatedData: T[][]
  totalPages: number
}
export const usePaginateData: <T>(data: T[], pageSize: number) => UsePaginateData<T> = (data, pageSize) => {
  const paginatedData = []
  const newEntries = [...data]
  while (newEntries.length > 0) {
    const paginatedSubData = newEntries.splice(0, pageSize)
    paginatedData.push(paginatedSubData)
  }
  return {
    paginatedData,
    totalPages: paginatedData.length,
  }
}

const useStyles = makeStyles(() => {
  return {
    paginationContainer: {
      display: 'flex',
      width: '100%',
      justifyContent: 'flex-end',
      alignItems: 'center',
    },
  }
})

interface PaginateData<T> {
  data: T[]
  pageSize: number
}

export interface PaginationProps<T> extends PaginateData<T> {
  handlePage: (page: T[]) => void
}

const Pagination: <T>(props: PropsWithChildren<PaginationProps<T>>) => ReactElement = ({
  data,
  pageSize,
  handlePage,
}) => {
  if (data.length === 0) {
    return <div aria-label="no-pagination-data" />
  }

  const { paginationContainer } = useStyles()
  const [page, setPage] = useState(0)
  const { paginatedData, totalPages } = usePaginateData<typeof data[0]>(data, pageSize)
  const visibleRows = `${page * pageSize + 1} - ${page * pageSize + paginatedData[page]?.length}`

  useEffect(() => {
    handlePage(paginatedData[0])
    setPage(0)
  }, [JSON.stringify(data)])

  const onPageChange = (newPage: number) => {
    setPage(newPage)
    handlePage(paginatedData[newPage])
    console.log('hit', newPage)
  }

  return (
    <div className={paginationContainer}>
      <span style={{ color: '#ababab', fontWeight: 700, marginRight: '8px' }}>
        {visibleRows} of {data.length}
      </span>
      <IconButton
        color="primary"
        aria-label="chevron-left"
        component="button"
        disabled={page === 0}
        onClick={() => onPageChange(page - 1)}
      >
        <ChevronLeft />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="chevron-right"
        component="button"
        disabled={page === totalPages - 1}
        onClick={() => onPageChange(page + 1)}
      >
        <ChevronRight />
      </IconButton>
    </div>
  )
}

export default Pagination
