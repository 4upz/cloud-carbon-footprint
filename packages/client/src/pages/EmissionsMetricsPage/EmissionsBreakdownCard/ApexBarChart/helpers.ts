/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { EmissionRatioResult } from '@cloud-carbon-footprint/common'
import { barChartCustomColors, Page, PageEntry } from 'Types'

const createCustomBarColors = (
  pageData: Page<PageEntry>,
  emissionsData: EmissionRatioResult[],
  mainTheme: string,
): string[] => {
  const regionColorsMap: string[] = []
  pageData.data.forEach((region) => {
    const currentRegion = region.x[0]
    let color = barChartCustomColors[0]
    const regionEmissionData = emissionsData.find(
      (item) => item.region === currentRegion,
    ) as EmissionRatioResult
    if (!regionEmissionData) {
      regionColorsMap.push(mainTheme)
    } else {
      const { mtPerKwHour } = regionEmissionData
      if (mtPerKwHour >= 0.00064) {
        color = barChartCustomColors[4]
      } else if (mtPerKwHour >= 0.00048 && mtPerKwHour < 0.00064) {
        color = barChartCustomColors[3]
      } else if (mtPerKwHour >= 0.00032 && mtPerKwHour < 0.00048) {
        color = barChartCustomColors[2]
      } else if (mtPerKwHour >= 0.00016 && mtPerKwHour < 0.00032) {
        color = barChartCustomColors[1]
      } else {
        color = barChartCustomColors[0]
      }
      regionColorsMap.push(color)
    }
  })
  return regionColorsMap
}

const mapToRange = (
  value: number,
  in_min: number,
  in_max: number,
  out_min: number,
  out_max: number,
): number => {
  return ((value - in_min) * (out_max - out_min)) / (in_max - in_min) + out_min
}

export { createCustomBarColors, mapToRange }
