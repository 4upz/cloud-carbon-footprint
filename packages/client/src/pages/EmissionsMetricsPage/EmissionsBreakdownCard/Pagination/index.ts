/*
 * Â© 2021 ThoughtWorks, Inc.
 */

export { default } from './Pagination'
export type { Page } from './Pagination'
