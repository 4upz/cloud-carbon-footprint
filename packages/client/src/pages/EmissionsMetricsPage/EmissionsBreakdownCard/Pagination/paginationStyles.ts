/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import { makeStyles } from '@material-ui/core'

const useStyles = makeStyles(() => {
  return {
    paginationContainer: {
      paddingTop: '10px',
      display: 'flex',
      width: '100%',
      justifyContent: 'flex-end',
      alignItems: 'center',
    },
    paginationLabel: {
      color: '#ababab',
      fontWeight: 700,
      marginRight: '8px',
    },
  }
})

export default useStyles
