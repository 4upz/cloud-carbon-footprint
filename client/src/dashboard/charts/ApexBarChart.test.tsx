/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { create, ReactTestRenderer } from 'react-test-renderer'
import { EstimationResult } from '../../models/types'
import moment from 'moment'
import React from 'react'
import { ApexBarChart } from './ApexBarChart'
import Chart from 'react-apexcharts'
import Pagination from './Pagination'

describe('ApexBarChart', () => {
  let fixture: ReactTestRenderer

  beforeEach(() => {
    const data: EstimationResult[] = [
      {
        timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
        serviceEstimates: [
          {
            cloudProvider: 'AWS',
            accountName: 'some account',
            serviceName: 'ebs',
            wattHours: 0,
            co2e: 3015.014,
            cost: 0,
            region: 'us-west-2',
          },
          {
            cloudProvider: 'AWS',
            accountName: 'some account',
            serviceName: 's3',
            wattHours: 0,
            co2e: 1718.017,
            cost: 0,
            region: 'us-west-2',
          },
          {
            cloudProvider: 'AWS',
            accountName: 'some account',
            serviceName: 'ec2',
            wattHours: 0,
            co2e: 2521.406,
            cost: 0,
            region: 'us-west-2',
          },
        ],
      },
    ]
    fixture = create(<ApexBarChart data={data} dataType="service" />)
  })
  it('renders with correct configuration', () => {
    expect(fixture.toJSON()).toMatchSnapshot()
  })

  it('should set the tooltip value formatter correctly', () => {
    const yFormatter = fixture.root.findByType(Chart).props?.options?.tooltip?.y?.formatter

    expect(yFormatter).toBeDefined()
    expect(yFormatter(1000.23)).toEqual('1000.230 mt')
  })

  it('should filterr, sort and order data and pass to Pagination component', function () {
    const paginationComponent = fixture.root.findByType(Pagination)
    const chartComponent = fixture.root.findByType(Chart)
    const chartOptionSeries = chartComponent.props?.options?.series[0]

    const sortedData = [
      { x: 'ebs', y: 3015.014 },
      { x: 'ec2', y: 2521.406 },
      { x: 's3', y: 1718.017 },
    ]

    expect(chartOptionSeries.data).toEqual([])
    expect(paginationComponent.props.data).toEqual(sortedData)
  })
})
