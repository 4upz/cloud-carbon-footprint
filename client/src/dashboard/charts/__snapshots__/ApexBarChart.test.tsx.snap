// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexBarChart renders with correct configuration 1`] = `
<div>
  <div
    data-testid="fake-bar-chart"
    height="500px"
    options={
      Object {
        "chart": Object {
          "toolbar": Object {
            "tools": Object {
              "download": null,
            },
          },
          "type": "bar",
        },
        "colors": Array [
          "#3f51b5",
        ],
        "dataLabels": Object {
          "background": Object {
            "borderColor": "#303f9f",
            "borderRadius": 1,
            "borderWidth": 1,
            "enabled": true,
            "foreColor": "#3f51b5",
            "opacity": 0.9,
            "padding": 6,
          },
          "enabled": true,
          "formatter": [Function],
          "offsetX": 10,
          "textAnchor": "start",
        },
        "grid": Object {
          "padding": Object {
            "left": 32,
          },
          "show": false,
          "xaxis": Object {
            "lines": Object {
              "show": false,
            },
          },
          "yaxis": Object {
            "lines": Object {
              "show": false,
            },
          },
        },
        "height": "500px",
        "plotOptions": Object {
          "bar": Object {
            "barHeight": "0%",
            "horizontal": true,
          },
        },
        "series": Array [
          Object {
            "data": Array [],
            "name": "Total CO2e",
          },
        ],
        "tooltip": Object {
          "fillSeriesColor": false,
          "x": Object {
            "show": false,
          },
          "y": Object {
            "formatter": [Function],
          },
        },
        "xaxis": Object {
          "axisBorder": Object {
            "show": false,
          },
          "labels": Object {
            "show": false,
          },
          "max": 3015.014,
          "type": "category",
        },
        "yaxis": Object {
          "labels": Object {
            "style": Object {
              "fontSize": "13px",
            },
          },
        },
      }
    }
    series={
      Array [
        Object {
          "data": Array [],
          "name": "Total CO2e",
        },
      ]
    }
    type="bar"
  >
    fake chart
  </div>
  <div
    className="makeStyles-paginationContainer-1"
  >
    <span
      style={
        Object {
          "color": "#ababab",
          "fontWeight": 700,
          "marginRight": "8px",
        }
      }
    >
      1 - 3
       of 
      3
    </span>
    <button
      aria-label="chevron-left"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-right"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
